[master] Acme fixes

jeffclay opened the pull request:
### What does this PR do?
#### Acme State
Added `dns_plugin_propogate_seconds` option to cert().
Added option to replace an existing staging/test certificate with a production cert.

#### Acme Module
Added `dns_plugin_propogate_seconds` option to cert(). This was previously in the doc string but not actually coded anywhere.
Added revoke() with optional reason.
Added delete().

### What issues does this PR fix or reference?
Fixes:
https://github.com/saltstack/salt/issues/63700
https://github.com/saltstack/salt/issues/64686

### Previous Behavior

1. `dns_plugin_propogate_seconds` only existed in doc strings.
2. After creating a certificate using `test_cert` flag then removing the flag, the result was nothing happening due to:
   1.  Acme state does check to see if cert exists and if it does it just checks if it needs renewed. No consideration for replacing a staging certificate. https://github.com/saltstack/salt/blob/75269c4ae22cce7cef1e681e2b1625368d6e18f5/salt/states/acme.py#L104 
   2. Acme module did not have capability to revoke or delete a certificate.

### New Behavior

1. `dns_plugin_propogate_seconds`  exists in Acme state and module
2. Acme state has new bool option `replace_staging` that will revoke and delete a staging certificate as long as "STAGING" is in the cert issuer dictionary and `test_cert` is no longer requested.
3. Acme module has new `revoke` capability which will delete a cert after revocation due to the `--non-interactive` cmd switch.
4. Acme module has new `delete` capability which deletes certificates. (No need to call this after revoke in sls state or Acme state, just seemed like a natural thing to add along with `revoke`.)

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [X] Docs
- [X] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

jeffclay committed:
Add propagation delay and stage certificate replacement options

Two new parameters were added to the ACME salt state to enhance configuration flexibility. The first, 'dns_plugin_propagate_seconds', allows specifying a delay for DNS propagation. The second, 'replace_staging', enables replacement of a same-name staging (test) certificate with a production one.

jeffclay committed:
Add revoke and delete functions in acme module

This commit introduces two new functions, "revoke" and "delete", to the ACME module, empowering users with the ability to revoke and delete certificates. An additional optional argument, "propagation_seconds", is also added for the Cloudflare DNS plugin which affects certificate renewal process.

jeffclay committed:
Add changelogs for PR

jeffclay commented:
This is a retry for [PR  #64685](https://github.com/saltstack/salt/pull/64685) since I mangled the re-base.

twangboy commented on salt/modules/acme.py:
```diff
@@ -421,3 +426,87 @@ def needs_renewal(name, window=None):
+
+
+def revoke(name, reason=None):
+    """
```
Needs a `.. version-added:: 3007` here

jeffclay replied:
I'm not sure I follow.

jeffclay replied:
@twangboy can you elaborate on what you're referring to?

twangboy replied:
When you add a new module, or a new function to a module, or a new parameter to a function, you need to note the version that that new item was added. That is done using `.. version-added:: <the version>`.

For example: https://github.com/saltstack/salt/blob/master/salt/modules/file.py#L597

jeffclay replied:
ah, ok. i hadn't seen that.

twangboy commented on salt/modules/acme.py:
```diff
@@ -421,3 +426,87 @@ def needs_renewal(name, window=None):
+
+
+def delete(name):
+    """
```
Needs a `.. version-added:: 3007` here

jeffclay committed:
Document version addition in 'revoke' and 'delete' functions

Added `versionadded` directive to the `revoke` and `delete` functions in the acme module

jeffclay committed:
Fixing ci/cd error:
/home/runner/work/salt/salt/salt/modules/acme.py:docstring of salt.modules.acme.delete:2: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/runner/work/salt/salt/salt/modules/acme.py:docstring of salt.modules.acme.revoke:2: WARNING: Explicit markup ends without a blank line; unexpected unindent.

jeffclay committed:
Add return statements in acme.py

The return statements have been added to the `delete` and `revocation` functions in the acme.py module. This ensures that each function will return a value of True upon successful completion

jeffclay committed:
Refactor acme module unit tests

The tests for the acme module have been simplified and streamlined for certificate revocation and deletion. The changes replace verbose setup dictionaries with more direct MagicMock patches and assertions, making the test cases more readable and maintainable. The code also better covers edge cases through more specific field checks in the assertion stage.

jeffclay commented:
@dwoz Do I need to do anything here?
